modules/gdpr_consent/gdpr_consent.module

```diff
@@ -0,0 +1,446 @@
+  }
+
+  return entity_load('gdpr_consent_agreement', $ids, $conditions, $reset);
+}
+
+/**
+ * Gets an array of all agreements, keyed by the machine name.
+ */
+function gdpr_consent_agreement_load_multiple_by_name($name = NULL) {
+  $signups = entity_load_multiple_by_name('gdpr_consent_agreement', isset($name) ? array($name) : FALSE);
+  return isset($name) ? reset($signups) : $signups;
+}
+
+/**
+ * Implements hook_field_info().
+ *
+ * Provides a user consent field type.
+ */
+function gdpr_consent_field_info() {
+  return array(
+    'gdpr_user_consent' => array(
+      'label' => t('GDPR Consent'),
+      'description' => t('Stores user consent for a particular agreement'),
+      'default_widget' => 'gdpr_consent_widget',
+      'default_formatter' => 'gdpr_consent_formatter',
+    ),
+  );
+}
+
+/**
+ * Implements hook_field_validate().
+ */
+function gdpr_consent_field_validate($entity_type, $entity, $field, $instance, $langcode, $items, &$errors) {
+  $valid = TRUE;
+
+  foreach ($items as $delta => $item) {
+    if (!empty($item['target_id'])) {
+      if (!$valid) {
+        $errors[$field['field_name']][$langcode][$delta][] = array(
+          'error' => 'gdpr_user_consent_invalid',
+          'message' => t('Referenced consent agreement entity is invalid.'),
+        );
+      }
+    }
+  }
+}
+
+/**
+ * Implements hook_field_is_empty().
+ */
+function gdpr_consent_field_is_empty($item, $field) {
+  return empty($item['target_id']);
+}
+
+/**
+ * Implements hook_field_formatter_info().
+ */
+function gdpr_consent_field_formatter_info() {
+  return array(
+    'gdpr_consent' => array(
+      'label' => t('GDPR user consent formatter'),
+      'field types' => array('gdpr_user_consent'),
+    ),
+  );
+}
+
+/**
+ * Implements hook_field_settings_form().
+ */
+function gdpr_consent_field_settings_form($field, $instance, $has_data) {
+  $settings = $field['settings'];
+
+  $agreements = array();
+
+  foreach (entity_load('gdpr_consent_agreement') as $agreement) {
+    $agreements[$agreement->id] = $agreement->title;
+  }
+
+  $form['target_id'] = array(
+    '#type' => 'select',
+    '#title' => t('User consent agreement'),
+    '#default_value' => (isset($settings['target_id'])) ? $settings['target_id'] : '',
+    '#required' => FALSE,
+    '#options' => $agreements,
+    '#element_validate' => array(
+      'element_validate_integer_positive',
+    ),
+    '#description' => t('The GDPR User Consent Agreement to display'),
+  );
+
+  return $form;
+}
+
+/**
+ * Implements hook_field_formatter_view().
+ */
+function gdpr_consent_field_formatter_view($entity_type, $entity, $field, $instance, $langcode, $items, $display) {
+  $element = array();
+
+  switch ($display['type']) {
+    case 'gdpr_consent':
+      foreach ($items as $delta => $item) {
+        $element[$delta] = array(
+          '#type' => 'html_tag',
+          '#tag' => 'p',
+          '#value' => t('User Consent ID: @entity', array('@entity' => $item['target_id'])),
```

Comment:
Do we want a second formatter that includes who registered the consent + notes etc?